rp2040-hal/src/gpio.rs

```diff
@@ -1,132 +1,337 @@
+/// The amount of current that a pin can drive when used as an output
+pub enum OutputDriveStrength {
+    /// 2 mA
+    Ma2,
+    /// 4 mA
+    Ma4,
+    /// 8 mA
+    Ma8,
+    /// 12 mA
+    Ma12,
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+/// The slew rate of a pin when used as an output
+pub enum OutputSlewRate {
+    /// Slew slow
+    Slow,
+    /// Slew fast
+    Fast,
+}
+
+// Magic numbers from the datasheet
+// const FUNCTION_SPI: u8 = 1;
+// const FUNCTION_UART: u8 = 2;
+// const FUNCTION_I2C: u8 = 3;
+// const FUNCTION_PWM: u8 = 4;
 const FUNCTION_SIO: u8 = 5;
+// const FUNCTION_PIO0: u8 = 6;
+// const FUNCTION_PIO1: u8 = 7;
+// const FUNCTION_CLOCK: u8 = 8;
+// const FUNCTION_USB: u8 = 9;
 
 macro_rules! gpio {
-    ($GPIOX:ident, $gpiox:ident, $PADSX:ident, [
-        $($PXi:ident: ($pxi:ident, $i:expr),)+
+    ($GPIOX:ident, $gpiox:ident, $PADSX:ident, $padsx:ident, $gpioxs:expr, [
+        $($PXi:ident: ($pxi:ident, $i:expr, $is:expr),)+
     ]) => {
-        impl GpioExt for pac::$GPIOX {
-            type Parts = Parts;
-
-            fn split(self, resets: &mut pac::RESETS) -> Parts {
-                resets.reset.modify(|_, w| w.$gpiox().clear_bit());
-                Parts {
-                    $(
-                        $pxi: $PXi { _mode: PhantomData },
-                    )+
+        #[doc = "HAL objects for the "]
+        #[doc = $gpioxs]
+        #[doc = " bank of GPIO pins"]
+        pub mod $gpiox {
+            use core::convert::Infallible;
+            use core::marker::PhantomData;
+            use embedded_hal::digital::v2::{InputPin, OutputPin, StatefulOutputPin};
+            use super::*;
+
+            impl GpioExt<pac::$PADSX, pac::SIO> for pac::$GPIOX {
+                type Parts = Parts;
+
+                fn split(self, pads: pac::$PADSX, sio: pac::SIO, resets: &mut pac::RESETS) -> Parts {
+                    resets.reset.modify(|_, w| w.$gpiox().clear_bit());
+                    Parts {
+                        _pads: pads,
+                        _sio: sio,
+                        $(
+                            $pxi: $PXi { _mode: PhantomData },
+                        )+
+                    }
                 }
             }
-        }
 
-        pub struct Parts {
-            $(
-                pub $pxi: $PXi<Input<Floating>>,
-            )+
-        }
+            #[doc = "Struct containing HAL objects for all the "]
+            #[doc = $gpioxs]
+            #[doc = " pins"]
+            pub struct Parts {
+                _pads: pac::$PADSX,
+                _sio: pac::SIO,
+                $(
+                    #[doc = "GPIO pin "]
+                    #[doc = $is]
+                    pub $pxi: $PXi<Unknown>,
+                )+
+            }
+
+            // Puts pad in same state as post-reset, according to datasheet
+            fn setup_pad_io(pads: &pac::$padsx::RegisterBlock, index: usize) {
+                pads.gpio[index].modify(|_, w| w.ie().set_bit()
```

Comment:
Sorry I did not notice this yesterday, but we can actually just use the [`.reset()`](https://docs.rs/rp2040-pac/0.1.1/rp2040_pac/generic/struct.Reg.html#method.reset) method to reset all the fields.